app/models/user.rb

```diff
@@ -11,11 +13,39 @@ class User < ApplicationRecord
     length: {maximum: Settings.user_model.email_maximun},
     format: {with: VALID_EMAIL_REGEX}, uniqueness: {case_sensitive: false}
   validates :password, presence: true,
-    length: {minimum: Settings.user_model.password_minximun}
-  validates :sex, inclusion: {in: :sex}
+    length: {minimum: Settings.user_model.password_minximun}, allow_nil: true
```

Comment:
cái allow_nil: true để làm gì?

Reply:
allow_nil cho phép cái trường đó được để trống, vì dùng update_attributes nó sẽ update các trường và validation, nên có nhg trường không có dữ liệu gửi lên, nên cần dùng cái đó

Reply:
Chưa rõ lắm a, vậy đăng ký tài khoản nó không cần trường này ạ?

Reply:
ở trương hợp đăng ký thì không cần nhưng ở trường hợp edit thì cần, với password,

Reply:
ý e là a bảo `cho phép cái trường đó được để trống` tức là lúc đăng ký nó để trống vẫn đc ạ?

Reply:
- Không, cái allow_nil cho cho phép cái trược đó được nil chứ không phải để trống, khi để trống nó vẫn báo lỗi valid. Trước a hiểu nhầm nil với để trống giống nhau, giờ a tìm hiểu lại rồi

Reply:
vậy nếu lúc gửi thông tin lên người dùng xóa cái field trong form rồi submit form thì field đó sẽ là nil hay trống?

Reply:
a chưa hiểu rõ câu hỏi đó, nên a sẽ trả lời theo nh trường 1.
1. Filed đó đang có dữ liệu (name) , người dùng xoá đi và sumbit thì đó là để trống.
2. Field đó trống sẵn, người dùng submit lên là để trống,
3. Có filed name và người vào code xoá field đó trong rồi submit lên thì sẽ là nil. ( tất nhiên là ở giao diện ng dùng thì k thể xoá dc field đó trong form của code)

Reply:
e F12 lên vẫn xóa được input và submit form bình thường mà.

Reply:
nếu dùng F12 thì sẽ là blank, vậy khi tạo đối tượng tất cả các trường đã có sẵn giá trị là nil, khhi submit dù không có thì nó vẫn gửi lên là blank. 
Chỉ ơ trường hợp update_attributes thì nó sẽ update cái trường ta muốn update, còn trường k muốn động vào nó sẽ gửi lên là nil.

Reply:
has_secure_password
Method này cần có gem BCrypt và yêu cầu trường password_digest để chạy.
Method sẽ tự động gọi validation cho mâtj khẩu.
Nó có include một module để thực hiện việc xác thực token với digest, tạo digest từ một mã đầu vào.

Reply:
gọi khi nào và gọi những validation gì cho mật khẩu a?

Reply:
Những validation : là present, lenght password (<=72), confir password..
has_secure_password được gọi khi quá trình validate đầu tiên của user được thực hiện

Reply:
e chưa hiểu rõ cái `quá trình validate đầu tiên` lắm, a nói rõ hơn cho e được không.
và ý câu hỏi trên kia của e là cái validate mà has_secure_password cung cấp đó sẽ được gọi khi nào?

Reply:
tức là khi submit form đăng ký, nó sẽ gọi tới validation mà mình đã đã định nghĩa trong model user, sau đó nó sẽ gọi tới hàm has_secure_password, và hàm ấy xử lý sau quá trình validation trên.
Vi singup như hôm qua, có 3 errorr lần lượt là 
Password can't be blank
Password is too short (minimum is 6 characters)
Password can't be blank
thì 2 cái đầu do validation đầu, và cái thứ 3 do validation trong has_secure_password

Reply:
vậy cũng vẫn để code như thế, lúc update a submit form thì còn bao nhiêu error? lý do có sự khác biệt đó là gì?

Reply:
lúc update thì sẽ có một lỗi, nêu để passwork trống. Lý do sự khác biệt theo a nghĩ là nó chỉ chạy validation ở has_secure_password.

Reply:
a đọc lại về has_secure_password kỹ hơn xíu nhé. ngay trong đoạn đó có nói rồi mà

Reply:
là has_secure_password được gọi trong quá trình create à

Reply:
(ok)